File: backend/arrlink/core/vocabulary.py
"""Vocabulary: known values per rule-condition category."""

from __future__ import annotations

import copy
import json
import os
from typing import Any

from ..state import State

TMDB_BASE = "https://api.themoviedb.org/3"

# Fallback certification country when no `tmdb_certification_country`
# Setting is configured (Settings > General > Region). Determines which
# country's certification scheme (e.g. PG-13 vs 12A/15/18) is fetched.
TMDB_CERTIFICATION_COUNTRY = "US"
DEFAULT_TMDB_API_KEY = os.environ.get("TMDB_API_KEY", "")
TRASH_QUALITY_URL = (
    "https://raw.githubusercontent.com/TRaSH-Guides/Guides/master/"
    "docs/json/radarr/quality-size/movie.json"
)

RICH_CATEGORIES = frozenset(
    {"genre", "certification", "collection", "quality", "language", "audio_language"}
)


def tmdb_api_key(db: State) -> str:
    """Effective TMDB key: a per-instance Settings override takes priority
    over the operator-configured default (env var); empty if neither set."""
    return (db.get_setting("tmdb_api_key") or "").strip() or DEFAULT_TMDB_API_KEY


async def _tmdb_get(path: str, api_key: str) -> Any:
    import httpx

    async with httpx.AsyncClient(timeout=15.0) as client:
        r = await client.get(
            f"{TMDB_BASE}{path}",
            params={"api_key": api_key, "language": "en"},
        )
    r.raise_for_status()
    return r.json()


async def sync_tmdb_vocabulary(db: State) -> dict[str, int]:
    """Fetch genre + certification lists from TMDB and persist them as
    shared (app_id NULL) vocabulary for both app types. No-ops (returns an
    empty dict) when no TMDB key is configured -- never raises, this is
    best-effort background refresh data, not core linking behavior."""
    key = tmdb_api_key(db)
    if not key:
        return {}

    counts: dict[str, int] = {}

    genre_paths = {"radarr": "/genre/movie/list", "sonarr": "/genre/tv/list"}
    for app_type, path in genre_paths.items():
        data = await _tmdb_get(path, key)
        genres = data.get("genres") if isinstance(data, dict) else None
        entries = [
            (str(g["name"]).strip(), str(g["id"]))
            for g in (genres or [])
            if isinstance(g, dict) and g.get("name")
        ]
        counts[f"genre:{app_type}"] = db.sync_vocabulary("genre", app_type, None, entries, "tmdb")

    cert_paths = {"radarr": "/certification/movie/list", "sonarr": "/certification/tv/list"}
    for app_type, path in cert_paths.items():
        data = await _tmdb_get(path, key)
        by_country = data.get("certifications") if isinstance(data, dict) else None
        country = db.get_setting("tmdb_certification_country") or TMDB_CERTIFICATION_COUNTRY
        rows = (by_country or {}).get(country) or []
        entries = [
            (str(c["certification"]).strip(), None)
            for c in rows
            if isinstance(c, dict) and c.get("certification")
        ]
        counts[f"certification:{app_type}"] = db.sync_vocabulary(
            "certification", app_type, None, entries, "tmdb"
        )

    return counts


async def sync_trash_vocabulary(db: State, app_type: str) -> int:
    """Fetch TRaSH Guides' quality-profile naming dictionary (public, no
    key) and persist as shared (app_id NULL) 'quality' vocabulary. Best
    effort -- logged and swallowed by the caller on failure."""
    import httpx

    async with httpx.AsyncClient(timeout=15.0) as client:
        r = await client.get(TRASH_QUALITY_URL)
    r.raise_for_status()
    data = r.json()
    names = {
        str(row["name"]).strip()
        for row in (data if isinstance(data, list) else [])
        if isinstance(row, dict) and row.get("name")
    }
    entries = [(name, None) for name in names]
    return db.sync_vocabulary("quality", app_type, None, entries, "trash")


def _vocabulary_values(db: State, category: str, app_type: str, app_ids: list[int]) -> set[str]:
    """Every known value for a scope: the shared (app_id IS NULL) vocabulary,
    plus every one of ``app_ids``' own instance-scoped vocabulary, and -- the
    tag classification bridge -- any tag on those apps manually classified
    into this category (treated as an equally-valid member of the category).
    ``app_ids`` is a single instance for an app-scoped rule, every enabled
    instance of that type for a type-scoped rule (see api/rules.py's
    _scope_app_ids), or [] for a real poll/preview against one known app."""
    if app_ids:
        placeholders = ",".join("?" for _ in app_ids)
        rows = db.query(
            f"SELECT value FROM vocabulary WHERE category=? AND app_type=? AND "
            f"(app_id IS NULL OR app_id IN ({placeholders}))",
            (category, app_type, *app_ids),
        )
    else:
        rows = db.query(
            "SELECT value FROM vocabulary WHERE category=? AND app_type=? AND app_id IS NULL",
            (category, app_type),
        )
    values = {r["value"] for r in rows}
    if app_ids:
        placeholders = ",".join("?" for _ in app_ids)
        tag_rows = db.query(
            f"SELECT label FROM tags WHERE app_id IN ({placeholders}) AND category=?",
            (*app_ids, category),
        )
        values |= {r["label"] for r in tag_rows}
    return values


def expand_vocabulary_conditions(
    rules: list[dict], db: State, app_id: int | None, app_type: str | None
) -> list[dict]:
    """Replace every match_type=='vocabulary' condition with an equivalent 'list' condition."""
    if app_type is None:
        # Unscoped rule (no specific app/app_type) -- nothing to resolve
        # against; leave any "vocabulary" condition as a no-op empty list
        # rather than guessing, or raising, mid-poll.
        expand_to = lambda cat: set()  # noqa: E731
    else:
        app_ids = [app_id] if app_id is not None else []
        expand_to = lambda cat: _vocabulary_values(db, cat, app_type, app_ids)  # noqa: E731

    from .planner import _rule_conditions  # local import: avoid a cycle (planner doesn't import us)

    out = []
    for rule in rules:
        r = copy.deepcopy(rule)
        conditions = copy.deepcopy(_rule_conditions(r))
        changed = False
        for cond in conditions:
            if cond.get("match_type") == "vocabulary":
                values = expand_to(cond.get("category"))
                cond["match_type"] = "list"
                cond["match_value"] = json.dumps(sorted(values))
                changed = True
        if changed:
            # _rule_conditions() prefers a pre-parsed "conditions" key over
            # conditions_json, so writing it back here is what makes the
            # expansion visible to match_conditions() downstream.
            r["conditions"] = conditions
        out.append(r)
    return out


def validate_condition_values(
    cond: dict, db: State, app_type: str | None, app_ids: list[int]
) -> list[str]:
    """Soft-warning check: for exact/list conditions in a category with a
    known vocabulary, flag literal values not found in that vocabulary and
    not covered by a tag manually classified into the category. Regex has
    no finite value to check, so it's never flagged. No-ops (returns [])
    when app_type is unresolved or the vocabulary scope is completely empty
    -- avoids a false "unknown" flood before anyone's synced anything.
    ``app_ids`` should be every instance in the rule's scope (see
    api/rules.py's _scope_app_ids), not just one representative instance --
    otherwise a value only a *different* instance of the same type happens
    to know about gets flagged as unknown."""
    category = cond.get("category")
    match_type = cond.get("match_type")
    if category not in RICH_CATEGORIES or match_type not in ("exact", "list"):
        return []
    if app_type is None:
        return []
    known = _vocabulary_values(db, category, app_type, app_ids)
    if not known:
        return []

    if match_type == "exact":
        candidates = [cond.get("match_value", "").strip()]
    else:
        candidates = [s.strip() for s in cond.get("match_value", "").split(",") if s.strip()]

    return [
        f"'{v}' is not a known {category} value for this app -- check spelling, "
        "or classify the matching tag into this category on the Tags page"
        for v in candidates
        if v and v not in known
    ]

File: backend-ts/src/core/vocabulary.ts
import { and, eq, inArray, isNull, or } from 'drizzle-orm'
import type { DbClient } from '../db/client.js'
import { tags as tagsTable, vocabulary } from '../db/schema.js'
import type { SettingsStore } from '../db/settings-store.js'
import type { Condition } from './matching.js'
import type { PlannerRule } from './planner.js'

/** Vocabulary: known values per rule-condition category. Ported from core/vocabulary.py. */

const TMDB_BASE = 'https://api.themoviedb.org/3'
const TMDB_CERTIFICATION_COUNTRY = 'US'
const TRASH_QUALITY_URL =
  'https://raw.githubusercontent.com/TRaSH-Guides/Guides/master/docs/json/radarr/quality-size/movie.json'

export const RICH_CATEGORIES = new Set([
  'genre',
  'certification',
  'collection',
  'quality',
  'language',
  'audio_language',
])

function isRecord(v: unknown): v is Record<string, unknown> {
  return typeof v === 'object' && v !== null
}

/** Effective TMDB key: a per-instance Setting overrides the operator-configured env default. */
export function tmdbApiKey(settingsStore: SettingsStore): string {
  const override = (settingsStore.getSetting<string>('tmdb_api_key') ?? '').trim()
  return override || process.env.TMDB_API_KEY || ''
}

async function tmdbGet(path: string, apiKey: string): Promise<unknown> {
  const url = new URL(`${TMDB_BASE}${path}`)
  url.searchParams.set('api_key', apiKey)
  url.searchParams.set('language', 'en')
  const r = await fetch(url)
  if (!r.ok) throw new Error(`TMDB HTTP ${r.status}`)
  return r.json()
}

/** Full-replace a (category, app_type, app_id) vocabulary scope. */
export function syncVocabularyRows(
  db: DbClient,
  category: string,
  appType: string,
  appId: number | null,
  entries: Array<[string, string | null]>,
  source: string,
): number {
  const now = Date.now() / 1000
  const scope = appId !== null ? eq(vocabulary.appId, appId) : isNull(vocabulary.appId)
  db.delete(vocabulary)
    .where(and(eq(vocabulary.category, category), eq(vocabulary.appType, appType), scope))
    .run()
  for (const [value, externalId] of entries) {
    db.insert(vocabulary)
      .values({ category, appType, appId, value, externalId, source, importedAt: now })
      .run()
  }
  return entries.length
}

/** Fetch genre + certification lists from TMDB, persisted as shared (app_id null)
 * vocabulary. No-ops when no TMDB key is configured -- best-effort, never throws for that case. */
export async function syncTmdbVocabulary(
  db: DbClient,
  settingsStore: SettingsStore,
): Promise<Record<string, number>> {
  const key = tmdbApiKey(settingsStore)
  if (!key) return {}

  const counts: Record<string, number> = {}

  const genrePaths: Record<string, string> = {
    radarr: '/genre/movie/list',
    sonarr: '/genre/tv/list',
  }
  for (const [appType, path] of Object.entries(genrePaths)) {
    const data = await tmdbGet(path, key)
    const genres = isRecord(data) && Array.isArray(data.genres) ? data.genres : []
    const entries: Array<[string, string]> = genres
      .filter((g): g is Record<string, unknown> => isRecord(g) && Boolean(g.name))
      .map((g) => [String(g.name).trim(), String(g.id)])
    counts[`genre:${appType}`] = syncVocabularyRows(
      db,
      'genre',
      appType,
      null,
      entries,
      'tmdb',
    )
  }

  const certPaths: Record<string, string> = {
    radarr: '/certification/movie/list',
    sonarr: '/certification/tv/list',
  }
  const country =
    settingsStore.getSetting<string>('tmdb_certification_country') ||
    TMDB_CERTIFICATION_COUNTRY
  for (const [appType, path] of Object.entries(certPaths)) {
    const data = await tmdbGet(path, key)
    const byCountry =
      isRecord(data) && isRecord(data.certifications) ? data.certifications : {}
    const rows = Array.isArray(byCountry[country]) ? byCountry[country] : []
    const entries: Array<[string, null]> = (rows as unknown[])
      .filter(
        (c): c is Record<string, unknown> => isRecord(c) && Boolean(c.certification),
      )
      .map((c) => [String(c.certification).trim(), null])
    counts[`certification:${appType}`] = syncVocabularyRows(
      db,
      'certification',
      appType,
      null,
      entries,
      'tmdb',
    )
  }

  return counts
}

/** Fetch TRaSH Guides' quality-profile naming dictionary (public, no key), persisted as
 * shared 'quality' vocabulary. Best-effort -- caller logs and swallows failures. */
export async function syncTrashVocabulary(
  db: DbClient,
  appType: string,
): Promise<number> {
  const r = await fetch(TRASH_QUALITY_URL)
  if (!r.ok) throw new Error(`TRaSH HTTP ${r.status}`)
  const data: unknown = await r.json()
  const names = new Set<string>()
  if (Array.isArray(data)) {
    for (const row of data as unknown[]) {
      if (isRecord(row) && typeof row.name === 'string' && row.name.trim()) {
        names.add(row.name.trim())
      }
    }
  }
  const entries: Array<[string, null]> = [...names].map((n) => [n, null])
  return syncVocabularyRows(db, 'quality', appType, null, entries, 'trash')
}

/** Every known value for a scope: shared (app_id null) vocabulary, plus every one of
 * `appIds`' own instance-scoped vocabulary and any tag manually classified into this
 * category. `appIds` is a single instance for an app-scoped rule, every enabled instance
 * of that type for a type-scoped rule (see api/rules.ts's scopeAppIds), or [] for a real
 * poll/preview against one already-known app (expandVocabularyConditions). */
function vocabularyValues(
  db: DbClient,
  category: string,
  appType: string,
  appIds: number[],
): Set<string> {
  const scope =
    appIds.length > 0
      ? or(isNull(vocabulary.appId), inArray(vocabulary.appId, appIds))
      : isNull(vocabulary.appId)
  const rows = db
    .select({ value: vocabulary.value })
    .from(vocabulary)
    .where(and(eq(vocabulary.category, category), eq(vocabulary.appType, appType), scope))
    .all()
  const values = new Set(rows.map((r) => r.value))
  if (appIds.length > 0) {
    const tagRows = db
      .select({ label: tagsTable.label })
      .from(tagsTable)
      .where(and(inArray(tagsTable.appId, appIds), eq(tagsTable.category, category)))
      .all()
    for (const r of tagRows) values.add(r.label)
  }
  return values
}

/** Replace every match_type=='vocabulary' condition with an equivalent 'list' condition. */
export function expandVocabularyConditions(
  rules: PlannerRule[],
  db: DbClient,
  appId: number | null,
  appType: string | null,
): PlannerRule[] {
  // Unscoped rule (no specific app/app_type): nothing to resolve against --
  // leave any vocabulary condition as a no-op empty list.
  const expandTo = (category: string): Set<string> =>
    appType === null
      ? new Set()
      : vocabularyValues(db, category, appType, appId !== null ? [appId] : [])

  return rules.map((rule) => ({
    ...rule,
    conditions: rule.conditions.map((cond): Condition => {
      if (cond.matchType !== 'vocabulary') return cond
      const values = expandTo(cond.category)
      return {
        ...cond,
        matchType: 'list',
        matchValue: JSON.stringify([...values].sort()),
      }
    }),
  }))
}

/** Soft-warning check: flags exact/list literal values not found in a category's known
 * vocabulary. Regex has no finite value to check, so it's never flagged. */
export function validateConditionValues(
  cond: Condition,
  db: DbClient,
  appType: string | null,
  appIds: number[],
): string[] {
  if (
    !RICH_CATEGORIES.has(cond.category) ||
    (cond.matchType !== 'exact' && cond.matchType !== 'list')
  ) {
    return []
  }
  if (appType === null) return []
  const known = vocabularyValues(db, cond.category, appType, appIds)
  if (known.size === 0) return []

  const candidates =
    cond.matchType === 'exact'
      ? [cond.matchValue.trim()]
      : cond.matchValue
          .split(',')
          .map((s) => s.trim())
          .filter(Boolean)

  return candidates
    .filter((v) => v && !known.has(v))
    .map(
      (v) =>
        `'${v}' is not a known ${cond.category} value for this app -- check spelling, ` +
        'or classify the matching tag into this category on the Tags page',
    )
}
